packages/contracts/contracts/SmartInvoiceFactoryV2.sol

```diff
@@ -0,0 +1,312 @@
+
+    event LogNewInvoice(
+        uint256 indexed index,
+        address invoice,
+        uint256[] amounts
+    );
+    event UpdateResolutionRate(
+        address indexed resolver,
+        uint256 indexed resolutionRate,
+        bytes32 details
+    );
+
+    address public immutable wrappedNativeToken;
+
+    constructor(address _wrappedNativeToken) {
+        require(
+            _wrappedNativeToken != address(0),
+            "invalid wrappedNativeToken"
+        );
+        wrappedNativeToken = _wrappedNativeToken;
+
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(ADMIN, msg.sender);
+    }
+
+    function _init(
+        address _invoiceAddress,
+        address _client,
+        address _provider,
+        uint8 _resolverType,
+        address _resolver,
+        uint256[] calldata _amounts,
+        bytes calldata _implementationData,
+        bytes32 _implementationType,
+        uint256 _implementationVersion,
+        address _implementationAddress
+    ) internal {
+        uint256 resolutionRate = resolutionRates[_resolver];
+        if (resolutionRate == 0) {
+            resolutionRate = 20;
+        }
+
+        ISmartInvoiceV2(_invoiceAddress).init(
+            _client,
+            _provider,
+            _encodeResolutionData(_resolverType, _resolver, resolutionRate),
+            _amounts,
+            wrappedNativeToken,
+            _implementationData,
+            _encodeImplementationData(
+                _implementationType,
+                _implementationVersion,
+                _implementationAddress,
+                invoiceCount
+            )
+        );
+
+        uint256 invoiceId = invoiceCount;
+        _invoices[invoiceId] = _invoiceAddress;
+        invoiceCount += 1;
+
+        emit LogNewInvoice(invoiceId, _invoiceAddress, _amounts);
+    }
+
+    // ******************
+    // Encoding
+    // ******************
+
+    function _encodeResolutionData(
+        uint8 _resolverType,
+        address _resolver,
+        uint256 _resolutionRate
+    ) internal pure returns (bytes memory) {
+        return abi.encode(_resolverType, _resolver, _resolutionRate);
+    }
+
+    function _encodeImplementationData(
+        bytes32 _implementationType,
+        uint256 _implementationVersion,
+        address _implementationAddress,
+        uint256 _invoiceId
+    ) internal pure returns (bytes memory) {
+        return
+            abi.encode(
+                _implementationType,
+                _implementationVersion,
+                _implementationAddress,
+                _invoiceId
+            );
+    }
+
+    // ******************
+    // Create
+    // ******************
+
+    function create(
+        address _client,
+        address _provider,
+        uint8 _resolverType,
+        address _resolver,
+        uint256[] calldata _amounts,
+        bytes calldata _implementationData,
+        bytes32 _implementationType
```

Comment:
We should work on better naming for these

Reply:
suggestions?

Reply:
sounds good for naming.

I'm not sure we need the extra param? That's the point of encoding implementationData as `bytes` right now; it can take a flexible number of arguments based on the implementation/invoice type.

Reply:
I've noticed the number of values you can decode can be limited and plus it should help with organization. Ideally, we should have only have these params:
- `_recipient` - just a clearer name for `provider`
- `_amounts` - only because our frontend is tightly coupled to this in the LogNewInvoice event subgraph. Otherwise I would encode this into the next param (`_invoiceData`)
- `_invoiceData` - main parameters for invoicing (including `client`) all encoded down into a single `bytes` param
- `_resolutionData` - any values needed for arbitration if applicable depending on invoice type
- `_ctx` - this is only used if future invoice types (like subscription or something else) needs additional or more arbitrary data that cannot be passed into `_invoiceData`. For our current implementations (escrow and pay now), this would just be empty `bytes`. But this allows us much greater flexibility for future types of implementations.

These should be lean and dynamic enough so we will able to serve all (or almost all) use cases for now and in the future.

Reply:
The issue with encoding is that there is definitely a gas cost incurred. I mentioned it above, but given that the resolver is needed to get the resolution rate, we would have to decode `_resolutionData` to get the resolver and resolution rate from the factory, then re-encode to pass it on to the invoice. To decode and re-encode together was about ~50k in gas when I tried it on remix w/ these three params. So if that's the route that we're going to take it just needs to be clear that the merits outweigh the drawbacks, which I'm not sure about yet.

What's the limit with encode/decode? I tried it with 10 strings & mixed types as well as a page of text on remix (with just one param) and worked on remix? It was super expensive but it got through. The only time it broke was when I tried 5 different pages at the same time, a string each, which would never happen.

For the above reasons and readability, I think we should leave the client and provider in at the top unless there's a compelling reason not to.

I like the idea of an empty extensible param as it could simplify things on the frontend. Again I haven't been able to hit a limit encoding/decoding anything less than 5 full pages of a string (albeit on a standalone contract). The original idea with `_invoiceData` is that it would handle this flexibility. As invoiceData could change with implementations I don't know if we should add another flexible param.

I like all the above ideas as it makes things flexible, but I just want to be aware of the cost factor as well the readability factor if it's too(*) stripped down (i.e. splitting up client and provider which will very likely be on every invoice).

Reply:
I don't know the exact number but that's why I originally suggested `_ctx` as another bytes param that could be decoded at a later period if needed in future implementations. However, I realized if we need to pass in so many params to require that, then we have failed to optimize our contracts. There are also possible workarounds like structs and arrays if necessary. And I was not suggesting we decode `_resolutionData` within factory, re-encode to pass through to only decode again within `init()`. That would defeat the whole purpose. I was suggesting we don't decode at all until its already passed through `init()` and decode within the implementation contract. #111 shows what I meant. I also optimized it further by scrapping `_resolutionData` and `_ctx`. All params besides provider and amounts are encoded in a single `bytes` param.

I can see implementations in which there is no single client such as a subscription/streaming contract that can accept payments from multiple customers. `_recipient` would be the exact same as `_provider` for the case of SmartInvoiceEscrow. However, I share an explanation for the name change in PR #111.